src/svg/qsvggenerator.cpp

```diff
@@ -997,8 +1007,37 @@ void QSvgPaintEngine::updateState(const QPaintEngineState &state)
     // close old state and start a new one...
     if (d->afterFirstUpdate)
         *d->stream << "</g>\n\n";
+    
+    QPainter* p = painter();
+    if (p->hasClipping()) {
+        QPainterPath painter_path = p->clipPathF();
+
+        if (painter_path.elementCount() > 0) {
+            QString clip_path = createPath(painter_path);
+
+            bool path_does_not_exist = 0 == clip_path_to_id.count(clip_path);
+            if (path_does_not_exist) {
+                clip_path_to_id[clip_path] = clip_counter++;
+                
+                d->stream->setString(&d->defs);
+
+                *d->stream << "<clipPath id=\"clip" << clip_path_to_id[clip_path] << "\">" << '\n';
+                *d->stream << '\t' << clip_path;
+                *d->stream << "</clipPath>" << '\n';
+
+                d->stream->setString(&d->body);
+            }
+
+            *d->stream << "<g clip-path=\"url(#clip" << clip_path_to_id[clip_path] << ")\" ";
+        }
+        else {
+            *d->stream << "<g ";
+        }
+    }
+    else {
+        *d->stream << "<g ";
+    }
```

Comment:
You could reduce this nesting by having a look at the existing implementation of this module. Note that it always streams the `<g` bit, and then streams various attributes within the tag such as brush setting etc.

So rather than having this nesting and repetition, always stream the tag opening, then if there is clipping stream the ` clip-path=\"url(#clip" << clip_path_to_id[clip_path] << ")\" ` bit only